library/VTKExtensions/Rendering/vtkF3DPointSplatMapper.cxx

```diff
@@ -0,0 +1,160 @@
+#include "vtkF3DPointSplatMapper.h"
+
+#include <vtkCamera.h>
+#include <vtkObjectFactory.h>
+#include <vtkOpenGLBufferObject.h>
+#include <vtkOpenGLIndexBufferObject.h>
+#include <vtkOpenGLPointGaussianMapperHelper.h>
+#include <vtkOpenGLRenderWindow.h>
+#include <vtkOpenGLRenderer.h>
+#include <vtkOpenGLShaderCache.h>
+#include <vtkOpenGLState.h>
+#include <vtkOpenGLVertexBufferObject.h>
+#include <vtkOpenGLVertexBufferObjectGroup.h>
+#include <vtkPolyData.h>
+#include <vtkShader.h>
+#include <vtkShaderProgram.h>
+#include <vtk_glew.h>
+
+#include "vtkF3DBitonicSort.h"
+#include "vtkF3DComputeDepthCS.h"
+
+//----------------------------------------------------------------------------
+class vtkF3DSplatMapperHelper : public vtkOpenGLPointGaussianMapperHelper
+{
+public:
+  static vtkF3DSplatMapperHelper* New();
+  vtkTypeMacro(vtkF3DSplatMapperHelper, vtkOpenGLPointGaussianMapperHelper);
+
+  vtkF3DSplatMapperHelper(const vtkF3DSplatMapperHelper&) = delete;
+  void operator=(const vtkF3DSplatMapperHelper&) = delete;
+
+protected:
+  vtkF3DSplatMapperHelper();
+
+  // overridden to create the OpenGL depth buffer
+  void BuildBufferObjects(vtkRenderer* ren, vtkActor* act) override;
+
+  // overridden to sort splats
+  void RenderPieceDraw(vtkRenderer* ren, vtkActor* act) override;
+
+private:
+  void SortSplats(vtkRenderer* ren);
+
+  vtkNew<vtkShader> DepthComputeShader;
+  vtkNew<vtkShaderProgram> DepthProgram;
+  vtkNew<vtkOpenGLBufferObject> DepthBuffer;
+
+  vtkNew<vtkF3DBitonicSort> Sorter;
+
+  double DirectionThreshold = 0.999;
+  double LastDirection[3];
+};
+
+//----------------------------------------------------------------------------
+vtkStandardNewMacro(vtkF3DSplatMapperHelper);
+
+//----------------------------------------------------------------------------
+vtkF3DSplatMapperHelper::vtkF3DSplatMapperHelper()
+{
+  this->LastDirection[0] = 0.0;
+  this->LastDirection[1] = 0.0;
+  this->LastDirection[2] = 0.0;
+
+  this->DepthComputeShader->SetType(vtkShader::Compute);
+  this->DepthComputeShader->SetSource(vtkF3DComputeDepthCS);
+  this->DepthProgram->SetComputeShader(this->DepthComputeShader);
+
+  this->Sorter->Initialize(512, VTK_FLOAT, VTK_UNSIGNED_INT);
+}
+
+//----------------------------------------------------------------------------
+void vtkF3DSplatMapperHelper::BuildBufferObjects(vtkRenderer* ren, vtkActor* act)
+{
+  vtkPolyData* poly = this->CurrentInput;
+
+  if (poly == nullptr)
+  {
+    return;
```

Comment:
should be testable

Reply:
Not really I think. I can convert it to an assert if you want